### PR TITLE
core: migrate stdcm and pathfinding modules to kotlin

### LINKED ISSUE #5751: core: migrate stdcm and pathfinding to kotlin
To do:

- [x] Automatically convert java to kotlin
- [x] Fix types in places where proper typing was missing (e.g. `Int` -> `BlockId`)
- [x] Fix any leftover compiling error
- [x] Fix any failed test
- [x] Double check for java function that haven't been replaced (e.g. `List.of`)
- [ ] Double check typing (`Distance / Length / Offset`)
- [x] Run code formatter

### PR DESCRIPTION
**Pull request by eckter**
To do:

- [x] Automatically convert java to kotlin
- [x] Fix types in places where proper typing was missing (e.g. `Int` -> `BlockId`)
- [x] Fix any leftover compiling error
- [x] Fix any failed test
- [x] Double check for java function that haven't been replaced (e.g. `List.of`)
- [ ] ~~Double check typing (`Distance / Length / Offset`)~~ edit: will be a different PR
- [x] Run code formater
- [x] Remove unnecessary `@JVMName`
- [x] Fix conflicts with main branch

### CODE REVIEW

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathPropUtils.kt [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+import fr.sncf.osrd.sim_infra.impl.buildChunkPath
+import fr.sncf.osrd.utils.Direction
+import fr.sncf.osrd.utils.indexing.DirStaticIdx
+import fr.sncf.osrd.utils.indexing.MutableDirStaticIdxArrayList
+import fr.sncf.osrd.utils.units.Distance.Companion.fromMeters
+import fr.sncf.osrd.utils.units.Length
+import fr.sncf.osrd.utils.units.Offset
+import fr.sncf.osrd.utils.units.meters
+
+object PathPropUtils {
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathPropUtils.kt]: Might be worth to look into moving some of these into `PathProperties`

Comment (multun) [reply core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathPropUtils.kt]: into the file or into the class? (also, `object PathPropUtils` creates a singleton, which isn't needed. These methods can just be moved out of the brackets)

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/GeomUtils.kt [hunk 2]
```diff
@@ -0,0 +1,14 @@
+package fr.sncf.osrd.api.pathfinding
+
+import fr.sncf.osrd.geom.LineString
+import fr.sncf.osrd.railjson.schema.geom.RJSLineString
+
+object GeomUtils {
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/GeomUtils.kt]: No need to define an object, probably java to kotlin automatism, will check for other occurences.

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.kt [hunk 3]
```diff
@@ -0,0 +1,408 @@
...
+            // TODO: include warnings in the response
+            ExceptionHandler.handle(ex)
+        }
+    }
+
+    /** Returns the list of dir track id on the route. Can contain duplicates.  */
+    private fun getRouteDirTracks(rawInfra: RawSignalingInfra, routeId: RouteId): DirStaticIdxList<TrackSection> {
+        val res = MutableDirStaticIdxArrayList<TrackSection>()
+        for (dirChunk in rawInfra.getChunksOnRoute(routeId)) {
+            val currentTrack: DirTrackSectionId = from(
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingBlocksEndpoint.kt]:
```suggestion
            val currentTrack = from(
```

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.kt [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+        rawPath: Pathfinding.Result<BlockId>
+    ): Collection<PathWaypointResult> {
+        // Builds a mapping between blocks and all user defined waypoints on the block
+        val userDefinedWaypointsPerBlock = HashMap<BlockId, MutableList<Distance>>()
+        for (waypoint in rawPath.waypoints) {
+            val offsets = userDefinedWaypointsPerBlock.computeIfAbsent(waypoint.edge) { _ -> ArrayList() }
+            offsets.add(waypoint.offset)
+        }
+        val res = ArrayList<PathWaypointResult>()
+        var lengthPrevBlocks: Distance = 0.meters
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.kt]:
```suggestion
        var lengthPrevBlocks = 0.meters
```

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.kt [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+import fr.sncf.osrd.graph.Pathfinding
+import fr.sncf.osrd.graph.Pathfinding.EdgeRange
+import fr.sncf.osrd.utils.indexing.*
+import fr.sncf.osrd.utils.units.Distance
+import fr.sncf.osrd.utils.units.meters
+import java.util.function.BiFunction
+import java.util.stream.Stream
+import kotlin.math.abs
+
+object PathfindingResultConverter {
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathfindingResultConverter.kt]: General comment here: couple of `Distance` values should be `Offset<ObjectId>` values.

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/RemainingDistanceEstimator.kt [hunk 6]
```diff
@@ -0,0 +1,88 @@
...
+import fr.sncf.osrd.graph.Pathfinding.EdgeLocation
+import fr.sncf.osrd.sim_infra.api.BlockId
+import fr.sncf.osrd.sim_infra.api.BlockInfra
+import fr.sncf.osrd.sim_infra.api.RawSignalingInfra
+import fr.sncf.osrd.utils.units.Distance
+import fr.sncf.osrd.utils.units.Distance.Companion.fromMeters
+
+/** This is a function object that estimates the remaining distance to the closest target point,
+ * using geo data. It is used as heuristic for A*.  */
+class RemainingDistanceEstimator(
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/RemainingDistanceEstimator.kt:14]: Same here: we'll finally be able to use `Offset<ObjectId>` :)

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/constraints/ElectrificationConstraints.kt [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+) : EdgeToRanges<BlockId> {
+    override fun apply(edge: BlockId): MutableCollection<Pathfinding.Range> {
+        val res = HashSet<Pathfinding.Range>()
+        val path = PathPropUtils.makePathProps(blockInfra, rawInfra, edge)
+        for (stock in rollingStocks)
+            res.addAll(getBlockedRanges(stock, path))
+        return res
+    }
+
+    companion object {
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/constraints/ElectrificationConstraints.kt:32]: Remove companion object, unneeded i believe.

File: core/src/main/kotlin/fr/sncf/osrd/api/stdcm/STDCMEndpoint.kt [hunk 8]
```diff
@@ -0,0 +1,155 @@
...
+                res.blocks, recorder
+            )
+            val response = STDCMResponse(simResult, pathfindingRes, res.departureTime)
+            RsJson(RsWithBody(STDCMResponse.adapter.toJson(response)))
+        } catch (ex: Throwable) {
+            ExceptionHandler.handle(ex)
+        }
+    }
+
+    companion object {
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/api/stdcm/STDCMEndpoint.kt]: Same here, do we really need a companion object ? (probably automatically created when switching to kt).

File: core/src/main/kotlin/fr/sncf/osrd/graph/EdgeRangeCost.kt [hunk 9]
```diff
@@ -0,0 +1,7 @@
+package fr.sncf.osrd.graph
+
+import fr.sncf.osrd.graph.Pathfinding.EdgeRange
+
+fun interface EdgeRangeCost<EdgeT> {
```
Comment (Erashin) [inline core/src/main/kotlin/fr/sncf/osrd/graph/EdgeRangeCost.kt]: Group the different `Edge` interface definitions (and maybe `Graph` as well) into the same file ?

**Review by Erashin — CHANGES_REQUESTED** [review]
Some of these changes can be treated in another pr, let me know.

File: core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathPropUtils.kt (hunk 1 above)
Comment (eckter) [reply core/src/main/kotlin/fr/sncf/osrd/api/pathfinding/PathPropUtils.kt]: Actually, there's a reason for this separation: moving it to `PathProperties` would add dependencies to `osrd.main` (which isn't something we'd want to do)

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt [hunk 10]
```diff
@@ -50,12 +47,11 @@ interface PathProperties {
...
 
 /** Build a Path from chunks and offsets, filtering the chunks outside the offsets */
-@JvmName("buildPathPropertiesFrom")
 fun buildPathPropertiesFrom(
     infra: TrackProperties,
     chunks: DirStaticIdxList<TrackChunk>,
-    pathBeginOffset: Distance,
-    pathEndOffset: Distance,
+    pathBeginOffset: Length<Block>,
+    pathEndOffset: Length<Block>,
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt:54]: I don't think that's right: from the standpoint of PathProperties, whether the distance from the first chunk is a block length is irrelevant. I think it should still be a distance

**Review by Erashin — APPROVED** [review]
Good job!

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/PathPropertiesImpl.kt [hunk 11]
```diff
@@ -246,26 +240,15 @@ fun getOffsetOfTrackLocationOnChunks(
...
-
 /** Build chunkPath, which is the subset of the given chunks corresponding to the beginOffset and endOffset. **/
-@JvmName("buildChunkPath")
 fun buildChunkPath(
     infra: TrackProperties,
     chunks: DirStaticIdxList<TrackChunk>,
-    pathBeginOffset: Distance,
-    pathEndOffset: Distance
+    pathBeginOffset: Length<Block>,
+    pathEndOffset: Length<Block>
```
Comment (multun) [inline core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/PathPropertiesImpl.kt:248]: Same, I don't think that's a block length. An offset in a track chunk maybe

**Review by multun — CHANGES_REQUESTED** [review]

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt (hunk 10 above)
Comment (eckter) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt:54]: will be fixed on [this other pr](https://github.com/osrd-project/osrd/pull/5830)

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/PathPropertiesImpl.kt (hunk 11 above)
Comment (multun) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/impl/PathPropertiesImpl.kt:248]: we'll fix it later

File: core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt (hunk 10 above)
Comment (multun) [reply core/kt-osrd-sim-infra/src/main/kotlin/fr/sncf/osrd/sim_infra/api/PathProperties.kt:54]: we'll fix it later

**Review by multun — APPROVED** [review]

**Pull request merged**